felix/nftables/match_builder.go

```diff
@@ -0,0 +1,356 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nftables
+
+import (
+	"fmt"
+	"math/bits"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/felix/generictables"
+	"github.com/projectcalico/calico/felix/proto"
+)
+
+var Wildcard = "*"
+
+var _ generictables.MatchCriteria = nftMatch{}
+
+// nftMatch implements the MatchCriteria interface for nftables.
+type nftMatch []string
+
+func Match() generictables.MatchCriteria {
+	return new(nftMatch)
+}
+
+func (m nftMatch) Render() string {
+	return strings.Join([]string(m), " ")
+}
+
+func (m nftMatch) String() string {
+	return fmt.Sprintf("MatchCriteria[%s]", m.Render())
+}
+
+func (m nftMatch) MarkClear(mark uint32) generictables.MatchCriteria {
+	if mark == 0 {
+		log.Panic("Probably bug: zero mark")
+	}
+	return append(m, fmt.Sprintf("meta mark & %#x == 0", mark))
+}
+
+func (m nftMatch) MarkNotClear(mark uint32) generictables.MatchCriteria {
+	if mark == 0 {
+		log.Panic("Probably bug: zero mark")
+	}
+	return append(m, fmt.Sprintf("meta mark & %#x != 0", mark))
+}
+
+func (m nftMatch) MarkSingleBitSet(mark uint32) generictables.MatchCriteria {
+	if bits.OnesCount32(mark) != 1 {
+		// Disallow multi-bit matches to force user to think about the mask they should use.
+		// For example, if you are storing a number in the mark then you likely want to match on its
+		// 0-bits too
+		log.WithField("mark", mark).Panic("MarkSingleBitSet() should only be used with a single mark bit")
+	}
+	return m.MarkMatchesWithMask(mark, mark)
+}
+
+func (m nftMatch) MarkMatchesWithMask(mark, mask uint32) generictables.MatchCriteria {
+	logCxt := log.WithFields(log.Fields{
+		"mark": mark,
+		"mask": mask,
+	})
+	if mask == 0 {
+		logCxt.Panic("Bug: mask is 0.")
+	}
+	if mark&mask != mark {
+		logCxt.Panic("Bug: mark is not contained in mask")
+	}
+	return append(m, fmt.Sprintf("meta mark & %#x == %#x", mask, mark))
+}
+
+func (m nftMatch) NotMarkMatchesWithMask(mark, mask uint32) generictables.MatchCriteria {
+	logCxt := log.WithFields(log.Fields{
+		"mark": mark,
+		"mask": mask,
+	})
+	if mask == 0 {
+		logCxt.Panic("Bug: mask is 0.")
+	}
+	if mark&mask != mark {
+		logCxt.Panic("Bug: mark is not contained in mask")
+	}
+	return append(m, fmt.Sprintf("meta mark & %#x != %#x", mask, mark))
+}
+
+func (m nftMatch) InInterface(ifaceMatch string) generictables.MatchCriteria {
+	return append(m, fmt.Sprintf("iifname %s", ifaceMatch))
+}
+
+func (m nftMatch) OutInterface(ifaceMatch string) generictables.MatchCriteria {
+	return append(m, fmt.Sprintf("oifname %s", ifaceMatch))
+}
+
+func (m nftMatch) RPFCheckPassed(acceptLocal bool) generictables.MatchCriteria {
+	// TODO: acceptLocal is not supported in nftables mode.
```

Comment:
As far as I've been able to discover...

e.g., the `iptables-translate` utility doesn't know what to do with it.